### DISCUSSION TITLE
Multi Tenant collections per tenant or documents per tenant ?

### DISCUSSION THREAD

**Discussion by psm-solution-group** [ravendb/ravendb#18023] [Q&A]
Hello,

I am building a multi tenant application and what to hear from you what would be the best in ravendb:

collections per tenant 
or
documents per tenant

I have seen a ravendb multi tenant article but want to hear here your opinion

**Comment by DejanMilicic** [comment]
The usual approach is to map all objects of one class into one collection. There is a default convention for this, so this approach works out of the box. Thinking about "collection per tenant" approach and its implications, I can project significant additional effort needed in that scenario. Can you please explain what makes "collection per tenant" an attractive option for you?

**Reply by pullmann4rent** [reply]
I think he mean that he use tenant/domains for each schema like this:

collections:
Tenant1.Users
Tenant2.Users
…

generally you have only one schema and store all data there 
Users and all data gets their

**Comment by ayende** [comment]
The options are basically:

1. Database per tenant (`db-northwing`, `db-southsand`, etc) 
2. Collection per tenant (`Users.Northwind`, `Users.Southsand`, etc)
3. `TenantId` field (`"TenantId": "Northwind"`)


I would usually go with 1 or 3, never 2. 

For 1 - that is useful when you have a tenants of significant size, and you want to be able to split them between nodes / clusters easily.
For 3 - that is preferrable when most of your tenants are small. That allows you to centralize resources far more efficiently.

**Reply by pullmann4rent** [reply]
i have 3 tenants but They have 1000 employees so is the #3 at the beginnend good for me?

and if I use the First Option how it manage the Connection if I get Every time new request from dB so I have to Switch everytime the Connection how I handle this in nodejs with Ravendb ?

**Reply by ItsRobbAllen** [reply]
We use 1 database per tenant and I wrote a library to handle the ability to switch databases for us. Using dependency injection, we have a session manager that inspects the running thread for identity information. 

With connections via our API, we use JWT and inspect the claim for their tenant and use that for their session. When we're responding to queue messages or other internal processes, we _create_ a claim provider using information we gather from various processes. We also have processes that run anonymously, and for those, the session manager takes in the name of the tenant directly.

It's complex(ish) in the setup, but now that it's in place, none of my devs concern themselves with their connections, it's automatic.

**Reply by pullmann4rent** [reply]
Thank you for the answer. Do you have an example how can I make to switching Connection?

**Reply by ayende** [reply]
RavenDB should handle that for you. You use the same document store and create a session with `openSession(dbName)`
